cdc/cdc.cc

```diff
@@ -218,4 +219,195 @@ db_context db_context::builder::build() {
     };
 }
 
+class transformer final {
+public:
+    struct map_cmp {
+        template <typename Func>
+        bool cmp(const std::pair<net::inet_address, unsigned int>& a,
+                 const std::pair<net::inet_address, unsigned int>& b,
+                 Func&& func) const {
+            auto&& a_ip = func(a.first).ip;
+            auto&& b_ip = func(b.first).ip;
+            return a_ip < b_ip || (a_ip == b_ip && a.second < b.second);
+        }
+        bool operator()(const std::pair<net::inet_address, unsigned int>& a,
+                        const std::pair<net::inet_address, unsigned int>& b) const {
+            if (a.first.is_ipv4()) {
+                return !b.first.is_ipv4() ||
+                    cmp(a, b, [] (const net::inet_address& addr) { return addr.as_ipv4_address(); });
+            } else {
+                return !b.first.is_ipv4() &&
+                    cmp(a, b, [] (const net::inet_address& addr) { return addr.as_ipv6_address(); });
+            }
+        }
+    };
+    using streams_type = std::map<std::pair<net::inet_address, unsigned int>, utils::UUID, map_cmp>;
+private:
+    db_context _ctx;
+    const schema& _schema;
+    schema_ptr _log_schema;
+    utils::UUID _time;
+    bytes _decomposed_time;
+    // The assumption here is that all nodes in the cluster are using the
+    // same murmur3_partitioner_ignore_msb_bits. If it's not true, we will
+    // have to gossip it around together with shard_count.
+    unsigned _ignore_msb_bits;
+    const streams_type& _streams;
+
+    clustering_key set_pk_columns(const partition_key& pk, int batch_no, mutation& m) const {
+        const auto log_ck = clustering_key::from_exploded(
+                *m.schema(), { _decomposed_time, int32_type->decompose(batch_no) });
+        auto pk_value = pk.explode(_schema);
+        size_t pos = 0;
+        for (const auto& column : _schema.partition_key_columns()) {
+            assert (pos < pk_value.size());
+            auto cdef = m.schema()->get_column_definition(to_bytes("_" + column.name()));
+            auto value = atomic_cell::make_live(*column.type,
+                                                _time.timestamp(),
+                                                bytes_view(pk_value[pos]));
+            m.set_cell(log_ck, *cdef, std::move(value));
+            ++pos;
+        }
+        return log_ck;
+    }
+    partition_key stream_id(const net::inet_address& ip, unsigned int shard_id) const {
+        auto it = _streams.find(std::make_pair(ip, shard_id));
+        if (it == std::end(_streams)) {
+                throw std::runtime_error(format("No stream found for node {} and shard {}", ip, shard_id));
```

Comment:
This is tricky, it will probably fail during topology changes. We should really move logging to the replica.

Reply:
The whole thing assumes there are no topology changes. And it will most certainly fail.
Please read my "CDC with topology change support" document, which is a modification to the original design. I'm working on the code right now.